core/src/main/java/com/linecorp/armeria/client/HttpClientDelegate.java

```diff
@@ -253,11 +286,24 @@ private static HttpResponse earlyFailedResponse(Throwable t, ClientRequestContex
         return HttpResponse.ofFailure(cause);
     }
 
+    private static HttpResponse earlyFailedResponse(Throwable t,
+                                                    ClientRequestContext ctx,
+                                                    DecodedHttpResponse response) {
+        final UnprocessedRequestException cause = UnprocessedRequestException.of(t);
+        ctx.cancel(cause);
+        response.close(cause);
+        return response;
+    }
+
     private static void doExecute(PooledChannel pooledChannel, ClientRequestContext ctx,
                                   HttpRequest req, DecodedHttpResponse res) {
         final Channel channel = pooledChannel.get();
         final HttpSession session = HttpSession.get(channel);
         res.init(session.inboundTrafficController());
         session.invoke(pooledChannel, ctx, req, res);
     }
+
+    private static InetSocketAddress createUnresolvedAddressForRefreshing(InetSocketAddress previousAddress) {
+        return new InetSocketAddress(previousAddress.getHostName(), previousAddress.getPort());
```

Comment:
`getHostName()` and the constructor may trigger a reverse lookup that is a blocking operation and can't reuse Armeria `DnsCache`.
```suggestion
        return InetSocketAddress.createUnresolved(previousAddress.getHostString(), previousAddress.getPort());
```